### PR TITLE
ASoC: SOF: intel: hda: Clean up link DMA for IPC3 during stop

### PR DESCRIPTION
**Pull request by ranj063**
With IPC3, we reset hw_params during the stop trigger, so we should also clean up the link DMA during the stop trigger.

Fixes: 1bf83fa6654c ("ASoC: SOF: Intel: hda-dai: Do not perform DMA cleanup during stop")

### CODE REVIEW

**Review by bardliao — DISMISSED** [review]

File: sound/soc/sof/intel/hda-loader.c [hunk 1]
```diff
@@ -549,6 +549,12 @@ int hda_dsp_ipc4_load_library(struct snd_sof_dev *sdev,
 	msg.primary |= SOF_IPC4_MSG_TYPE_SET(SOF_IPC4_GLB_LOAD_LIBRARY);
 	msg.primary |= SOF_IPC4_MSG_DIR(SOF_IPC4_MSG_REQUEST);
 	msg.primary |= SOF_IPC4_MSG_TARGET(SOF_IPC4_FW_GEN_MSG);
+
+	/* send LOAD_LIBRARY IPC with lib_id = 0 to allow the FW to configure the DMA */
+	ret = sof_ipc_tx_message_no_reply(sdev->ipc, &msg, 0);
+	if (ret < 0)
+		return ret;
```
Comment (lyakh) [inline sound/soc/sof/intel/hda-loader.c]: this doesn't belong to this PR, does it?

**Review by lyakh — CHANGES_REQUESTED** [review]

File: sound/soc/sof/intel/hda-loader.c (hunk 1 above)
Comment (ranj063) [reply sound/soc/sof/intel/hda-loader.c]: haha sorry bad branch!

File: sound/soc/sof/intel/hda-dai-ops.c [hunk 2]
```diff
@@ -524,16 +524,25 @@ static const struct hda_dai_widget_dma_ops hda_ipc4_chain_dma_ops = {
...
+		int ret;
 
 		data.dai_data = DMA_CHAN_INVALID;
-		return hda_dai_config(w, SOF_DAI_CONFIG_FLAGS_HW_FREE, &data);
+		ret = hda_dai_config(w, SOF_DAI_CONFIG_FLAGS_HW_FREE, &data);
+		if (ret < 0)
+			return ret;
+
+		if (cmd == SNDRV_PCM_TRIGGER_STOP)
+			return hda_link_dma_cleanup(substream, hext_stream, cpu_dai);
```
Comment (plbossart) [inline sound/soc/sof/intel/hda-dai-ops.c:543]: I don't understand how this is compatible with the previous direction to keep the stream tag in  https://github.com/thesofproject/linux/commit/1bf83fa6654ce8959948878aad14a6db586125b8 whose commit message included 

"In the case of repeated start/stop without involving hw_free, the stream
tag needs to be preserved for the subsequent starts. So, skip performing
the DMA clean up during stop and handle it only during suspend or
hw_free.
"

??

**Review by plbossart — COMMENTED** [review]
not following @ranj063

File: sound/soc/sof/intel/hda-dai-ops.c (hunk 2 above)
Comment (ranj063) [reply sound/soc/sof/intel/hda-dai-ops.c:543]: @plbossart that commit was a mistake for IPC3. In the case of IPC3, the post_trigger op frees clears the DMA channel ID and send the DAI_CONFIG to clear the DMA iD in the firmware as well. So we must clean up the link DMA set up so that we can re-assign and do the DAI_CONFIG again during prepare in the case of repeasted hw_params without hw_free.

Comment (plbossart) [reply sound/soc/sof/intel/hda-dai-ops.c:543]: ah ok, the commit https://github.com/thesofproject/linux/commit/1bf83fa6654ce8959948878aad14a6db586125b8 was for IPC4, makes sense now. Thanks for the clarification.

**Review by plbossart — APPROVED** [review]

**Comment by plbossart** [comment]
Don't we need additional tags @ranj063 

Closes: https://github.com/thesofproject/linux/issues/4455
Closes: https://github.com/thesofproject/linux/issues/4482

**Comment by ranj063** [comment]
> Don't we need additional tags @ranj063
> 
> Closes: #4455 Closes: #4482



> Don't we need additional tags @ranj063
> 
> Closes: #4455 Closes: #4482

update now. Thanks @plbossart

**Review by RanderWang — APPROVED** [review]

**Review by ujfalusi — APPROVED** [review]

**Review by lyakh — APPROVED** [review]
I guess an alternative way to fix https://github.com/thesofproject/linux/commit/1bf83fa6654ce8959948878aad14a6db586125b8 would be by reverting it for IPC3 - by re-adding that line and surrounding it with `#if CONFIG_...` / `#endif`

**Review by bardliao — APPROVED** [review]

**Comment by ranj063** [comment]
> I guess an alternative way to fix [1bf83fa](https://github.com/thesofproject/linux/commit/1bf83fa6654ce8959948878aad14a6db586125b8) would be by reverting it for IPC3 - by re-adding that line and surrounding it with `#if CONFIG_...` / `#endif`

@lyakh #if CONFIG_ wontwork because both IPC3 & IPC4 are always enabled by default
